### PR TITLE
add extend_lock for StorageLock

### PR DESCRIPTION
**Pull request by wangjj9219**
In some cases, such as doing iterations in offchain worker, the time to be consumed is hard to estimate
, and it's necessary to extend the deadline of the `StorageLock` after each successful next step during the whole iteration process.

So I add new function for `StorageLock` to support extending deadline of an active lock.

### CODE REVIEW

**Comment by drahnr** [comment]
Hi wanjgjj, thanks for your contribution - can you elaborate a bit on the usecase you have in mind? Could you outline one pseudo code example that would benefit from your proposed API?

**Comment by xlc** [comment]
We have built our own offchain storage lock before the one in Substrate is implemented. And now we would like to replace our one with the Substrate provided one. But we will need this feature to be included before we can replace it.

Our offchain lock:
https://github.com/AcalaNetwork/Acala/blob/mandala-tc3/utilities/src/offchain_lock.rs 

The use of extend lock:
https://github.com/AcalaNetwork/Acala/blob/c15a1910217810333daa480fa5e612a57c8a3281/modules/cdp_engine/src/lib.rs#L353

Basically we are iterating and process a potentially large amount of data in offchain worker. While a worker is processing the data, it will like to keep on hold the lock without have it expires in order to prevent other workers from been doing the same thing.

**Comment by drahnr** [comment]
The lock itself will unlock once the guard goes out of scope, unless you call `forget` on the guard.
The simple solution is to use a conservative timeout , since - in your case - iiuc the timeout is not really helping your case but rather intervening.
Setting the timeout one or two orders of magnitude higher than what you expect. But comes with a side effect: should your node ever be forcefully shut down, that lock would prevent the ocw execution for a considerable amount of time when restarting the node.

At the end of the day, I think this would be a good extension, @tomusdrw any thoughts?

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 1]
```diff
@@ -264,6 +264,24 @@ impl<'a, L: Lockable> StorageLock<'a, L> {
 		}
 	}
 
+	/// Extend ative lock's deadline
```
Comment (gavofyork) [inline primitives/runtime/src/offchain/storage_lock.rs:267]:
```suggestion
	/// Extend active lock's deadline
```

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 2]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
 	pub fn forget(mut self) {
 		let _ = self.lock.take();
 	}
+
+	/// Extend the lock by guard if the lock is existed.
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:359]:
```suggestion
	/// Extend the lock deadline iff it already exists.
```

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 3]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
...
+	///
+	/// Can be used to extend deadline when doing iterations
+	/// that it's hard to be estimated consumed time. After
+	/// a successful step, extend the lock if the lock is still active.
+	pub fn extend_lock(&mut self) {
+		if let Some(ref mut lock) = self.lock {
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:365]: I think this should return information whether extension was successful, so one can react on that.

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 4]
```diff
@@ -512,4 +541,47 @@ mod tests {
...
+			let mut lock = StorageLock::<'_, Time>::new(b"lock_4");
+			let res = lock.try_lock();
+			assert!(res.is_ok());
+			let guard = res.unwrap();
+			guard.forget();
+		});
+
+		// lock must have been cleared at this point
+		let opt = state.read().persistent_storage.get(b"", b"lock_4");
+		assert!(opt.is_some());
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:585]: Should we not probe for a value (or value range at least)?

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 5]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
 	pub fn forget(mut self) {
 		let _ = self.lock.take();
 	}
+
+	/// Extend the lock by guard if the lock is existed.
+	///
+	/// Can be used to extend deadline when doing iterations
+	/// that it's hard to be estimated consumed time. After
+	/// a successful step, extend the lock if the lock is still active.
+	pub fn extend_lock(&mut self) {
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:364]: The name `extend` is slightly misleading, since we really restart the deadline "countdown" so to speak, we don't prolong it by a value, but I don't have a better name either.

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 6]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
 	pub fn forget(mut self) {
 		let _ = self.lock.take();
 	}
+
+	/// Extend the lock by guard if the lock is existed.
+	///
+	/// Can be used to extend deadline when doing iterations
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:361]:
```suggestion
	/// Can be used to extend a deadline when having processing
	/// i.e. large sets of items for which it is hard to calculate a
	/// meaning full upper bound which does not block for a very
	/// long time on node termination.
```

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 7]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
...
+	/// that it's hard to be estimated consumed time. After
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:362]:
```suggestion
```

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 8]
```diff
@@ -337,6 +355,17 @@ impl<'a, 'b, L: Lockable> StorageLockGuard<'a, 'b, L> {
...
+	/// a successful step, extend the lock if the lock is still active.
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:363]:
```suggestion
```

File: primitives/runtime/src/offchain/storage_lock.rs (hunk 3 above)
Comment (tomusdrw) [reply primitives/runtime/src/offchain/storage_lock.rs:365]: Agree with this one. Currently you might be calling `extend_lock` in a loop, thinking that it's still acquired, while in reality it is just going to do nothing.
What's even worse you can have such sequence:
1. Lock is acquired by Worker 1
2. Lock reaches deadline, but Worker 1 still holds the `Guard`
3. Lock is acquired by Worker 2
4. Lock is extended by Worker 1 (but the owner technically is worker 2)

Of course this was an issue with previous API (that merely holding a `Guard` does not guarantee the lock being still active), but with `extend_lock` pattern I feel like we should give some extra guarantees, cause it suggests that it's okay to hold the lock for a longer periods of time and extend it from time to time (Without it was clear from the design that lock is meant to be held for some critical section that should complete fast).

File: primitives/runtime/src/offchain/storage_lock.rs (hunk 5 above)
Comment (wangjj9219) [reply primitives/runtime/src/offchain/storage_lock.rs]: Is `refresh` ok?

Comment (drahnr) [reply primitives/runtime/src/offchain/storage_lock.rs]: I think we can leave as is, I did not come up with a better name.

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 9]
```diff
@@ -264,6 +264,24 @@ impl<'a, L: Lockable> StorageLock<'a, L> {
...
 
+	/// Extend active lock's deadline
+	fn extend_active_lock(&mut self) -> Result<<L as Lockable>::Deadline, ()> {
+		let res = self.value_ref.mutate(|s: Option<Option<L::Deadline>>| -> Result<<L as Lockable>::Deadline, ()> {
+			match s {
+				// lock is present and is still active, extend the lock.
+				Some(Some(deadline)) if !<L as Lockable>::has_expired(&deadline) =>
+					Ok(self.lockable.deadline()),
+				// other cases
+				_ => Err(()),
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:275]: :+1:

File: primitives/runtime/src/offchain/storage_lock.rs [hunk 10]
```diff
@@ -512,4 +543,51 @@ mod tests {
...
+            let res = lock.try_lock();
+            assert!(res.is_ok());
+            let guard = res.unwrap();
+
+            guard.forget();
+        });
+
+        // lock must have been cleared at this point
+        let opt = state.read().persistent_storage.get(b"", b"lock_4");
+        assert_eq!(opt.unwrap(), vec![132_u8, 3u8, 0, 0, 0, 0, 0, 0]); // 132 + 256 * 3 = 900
```
Comment (drahnr) [inline primitives/runtime/src/offchain/storage_lock.rs:591]: :+1:
